src/System.Text.Utf8String/System/Text/Utf8Span.cs

```diff
@@ -0,0 +1,782 @@
+
+        static Utf8Span s_empty => default;
+
+        // TODO: Validate constructors, When should we copy? When should we just use the underlying array?
+        // TODO: Should we be immutable/readonly?
+        public Utf8Span(ReadOnlySpan<byte> buffer)
+        {
+            _buffer = buffer;
+        }
+
+        public Utf8Span(byte[] utf8bytes)
+        {
+            _buffer = new ReadOnlySpan<byte>(utf8bytes);
+        }
+
+        public Utf8Span(byte[] utf8bytes, int index, int length)
+        {
+            _buffer = new ReadOnlySpan<byte>(utf8bytes, index, length);
+        }
+
+        public Utf8Span(string s)
+        {
+            if (s == null)
+            {
+                throw new ArgumentNullException("s", "String cannot be null");
+            }
+
+            if (s == string.Empty)
+            {
+                _buffer = ReadOnlySpan<byte>.Empty;
+            }
+            else
+            {
+                _buffer = new ReadOnlySpan<byte>(GetUtf8BytesFromString(s));
+            }
+        }
+
+        /// <summary>
+        /// This constructor is for use by the compiler.
+        /// </summary>
+        [EditorBrowsable(EditorBrowsableState.Never)]
+        public Utf8Span(RuntimeFieldHandle utf8Data, int length) : this(CreateArrayFromFieldHandle(utf8Data, length))
+        {
+        }
+
+        public static explicit operator Utf8Span(ArraySegment<byte> utf8Bytes)
+        {
+            return new Utf8Span(utf8Bytes);
+        }
+
+        static byte[] CreateArrayFromFieldHandle(RuntimeFieldHandle utf8Data, int length)
+        {
+            var array = new byte[length];
+            RuntimeHelpers.InitializeArray(array, utf8Data);
+            return array;
+        }
+
+        public static Utf8Span Empty { get { return s_empty; } }
+
+        /// <summary>
+        /// Returns length of the string in UTF-8 code units (bytes)
+        /// </summary>
+        public int Length
+        {
+            get
+            {
+                return _buffer.Length;
+            }
+        }
+
+        public Enumerator GetEnumerator()
+        {
+            return new Enumerator(_buffer);
+        }
+
+        public CodePointEnumerable CodePoints
+        {
+            get
+            {
+                return new CodePointEnumerable(_buffer);
+            }
+        }
+
+        public byte this[int i]
+        {
+            [MethodImpl(MethodImplOptions.AggressiveInlining)]
+            get
+            {
+                // there is no need to check the boundaries -> Span is going to do this on it's own
+                return (byte)_buffer[i];
+            }
+        }
+
+        public static implicit operator ReadOnlySpan<byte>(Utf8Span utf8)
+        {
+            return utf8.Bytes;
+        }
+
+        public static explicit operator Utf8Span(string s)
+        {
+            return new Utf8Span(s);
+        }
+
+        public static explicit operator string(Utf8Span s)
```

Comment:
Maybe we should remove this. It will be more explicit and discoverable to simply call ToString

Reply:
Let's just remove this ctor.